java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -88,141 +94,182 @@ class EntityHandleManager(
+    private fun Set<RawEntity>.adaptValues() =
+        map { entitySpec.deserialize(it) }.toSet()
 
-    override suspend fun size(): Int {
-        checkCanRead()
-        return fetchAll().size
-    }
+    override suspend fun fetchAll() = storageHandle.fetchAll().adaptValues()
 
-    override suspend fun isEmpty(): Boolean {
-        checkCanRead()
-        return fetchAll().isEmpty()
-    }
-
-    override suspend fun fetchAll(): Set<T> {
-        checkCanRead()
-        return storageHandle.fetchAll().map {
-            entitySpec.deserialize(it)
-        }.toSet()
-    }
-
-    override suspend fun onUpdate(action: (Set<T>) -> Unit) {
-        checkCanRead()
-        storageHandle.addOnUpdate { raw ->
-            action(raw.map { entitySpec.deserialize(it) }.toSet())
-        }
+    override suspend fun onUpdate(action: (Set<T>) -> Unit) = storageHandle.addOnUpdate {
+        action(it.adaptValues())
     }
+}
 
+/** Implementation of collection write operations to mix into concrete instances. */
+private class WriteCollectionOperationsImpl<T : Entity>(
+    private val storageHandle: CollectionHandle<RawEntity>,
+    private val idGenerator: Id.Generator
+) : WriteCollectionOperations<T> {
     override suspend fun store(entity: T) {
-        checkCanWrite()
         storageHandle.store(
-            entity.ensureIdentified(idGenerator, name).serialize()
+            entity.ensureIdentified(idGenerator, storageHandle.name).serialize()
         )
     }
 
     override suspend fun clear() {
-        checkCanWrite()
         storageHandle.clear()
     }
 
     override suspend fun remove(entity: T) {
-        checkCanWrite()
         storageHandle.remove(entity.serialize())
     }
 }
 
+/** Base functionality common to all read/write singleton and collection handles. */
+abstract class BaseHandleAdapter(
+    private val storageHandle: StorageHandle<*, *, *>
+) : Handle {
+    // VisibleForTesting
+    val actorName = storageHandle.name
+
+    override val name = storageHandle.name
+
+    override suspend fun onSync(action: () -> Unit) = storageHandle.addOnSync(action)
+
+    override suspend fun onDesync(action: () -> Unit) = storageHandle.addOnDesync(action)
+}
+
+/** Delegate this interface in a concrete singleton handle impl to mixin read operations. */
+private interface ReadSingletonOperations<T : Entity> {
+    suspend fun fetch(): T?
+    suspend fun onUpdate(action: (T?) -> Unit)
+}
+
+/** Delegate this interface in a concrete singleton handle impl to mixin write operations. */
+private interface WriteSingletonOperations<T : Entity> {
+    suspend fun store(entity: T)
+    suspend fun clear()
+}
+
+/** Delegate this interface in a concrete collection handle impl to mixin read operations. */
+private interface ReadCollectionOperations<T : Entity> {
+    suspend fun size(): Int
+    suspend fun isEmpty(): Boolean
+    suspend fun fetchAll(): Set<T>
+    suspend fun onUpdate(action: (Set<T>) -> Unit)
```

Comment:
Technically you could move `onUpdate(action: (T) -> Unit)` onto a base interface like `UpdateHandle<T>`

Then you could make `ReadSingletonOperationsImpl<T> : UpdateHandle<T?>` and `WriteCollectionOperationsImpl<T> : UpdateHandle<Set<T>>`

BTW, we don't fire onSync() events for Write-only handles. I think this needs to be clarified, because if you write code that registers listeners on them, and expect it, they will never be called.

Reply:
(From OOB conversations, it sounds like you meant for the second sentence to say "and `ReadCollectionOerationsImpl...`") -- yes, I can make this change.

In the old impementation, we didn't fire `onSync` for write-only handles when attaching a new listener. After #4935, this is no longer true: onSync will be fired for both read and write handles, any time they attach an onSync listener, if the SP is synced.